Primary send safe commit in file-based recovery

dnhatn opened the pull request:
Today a primary shard transfers the most recent commit point to a replica shard in a file-based recovery. However, the most recent commit may not be a "safe" commit; this causes a replica shard not having a safe commit point until it can retain a safe commit by itself. 

This commits collapses the snapshot deletion policy into the combined deletion policy and modifies the peer recovery source to send a safe commit.

Relates #10708

bleskes commented on core/src/main/java/org/elasticsearch/index/engine/CombinedDeletionPolicy.java:
```diff
@@ -70,18 +76,51 @@ public void onInit(List<? extends IndexCommit> commits) throws IOException {
+        assert lastCommit != null : "Last commit is not initialized yet";
+        final IndexCommit snapshotting = acquiringSafeCommit ? safeCommit : lastCommit;
+        snapshottedCommits.addTo(snapshotting, 1); // increase refCount
+        return new Engine.IndexCommitRef(snapshotting, () -> releaseCommit(snapshotting));
```
can we leave this to the engine and make the releaseCommit method public?

bleskes replied:
we can also return a commit wrapper that disable the delete method?

dnhatn replied:
💯

bleskes commented on core/src/main/java/org/elasticsearch/index/engine/CombinedDeletionPolicy.java:
```diff
@@ -70,18 +76,51 @@ public void onInit(List<? extends IndexCommit> commits) throws IOException {
+        assert refCount >= 0 : "Number of snapshots can not be negative [" + refCount + "]";
+        if (refCount == 0) {
+            snapshottedCommits.remove(releasingCommit);
+            updateTranslogDeletionPolicy();
```
why does the translog care about the snapshotting situation? it only cares about the safe commit, no?

dnhatn replied:
Let's say we have two commits `c1` and `c2` with `c1` is the safe commit and `c2` is the last commit. Clients acquire `c1`, then we have to keep translog of `c1` until they release the commit. During that time, we have a new commit (or global checkpoint advanced), we can release `c1` and its translog but have to keep them as they are being snapshotted. When clients release the commit `c1`, we should also release its translog rather than wait until the next `onCommit`.

dnhatn replied:
I think we should not release translog here. We should either keep or release both the commit and translog at the same time. I will update this.

bleskes commented on core/src/main/java/org/elasticsearch/index/engine/Engine.java:
```diff
@@ -859,9 +857,10 @@ public void forceMerge(boolean flush) throws IOException {
      * @param flushFirst indicates whether the engine should flush before returning the snapshot
      */
-    public abstract IndexCommitRef acquireIndexCommit(boolean flushFirst) throws EngineException;
+    public abstract IndexCommitRef acquireIndexCommit(boolean safeCommit, boolean flushFirst) throws EngineException;
```
we should look in future to remove one of these booleans - either you want to "get everything" or you want a safe commit. I don't think there's a point in "flush but give me a safe commit"

dnhatn replied:
I agree. Are you ok if we replace this by having two methods: `acquireLastIndexCommit(flushFirst)` and `acquireSafeIndexCommit(no option)`?

bleskes replied:
>  agree. Are you ok if we replace this by having two methods: acquireLastIndexCommit(flushFirst) and acquireSafeIndexCommit(no option)?

+1

bleskes replied:
double checking - the new methods will be a follow up?

dnhatn commented:
@bleskes, I've updated the deletion policy to return a non-deletable commit and wrap it into a `commit-ref` in the engine. I am not happy with the casting line in the `releaseCommit` method `[final IndexCommit releasingCommit = ((SnapshotIndexCommit) snapshotCommit).delegate;]`. I will reach out to discuss this with you.

bleskes commented on core/src/main/java/org/elasticsearch/index/engine/CombinedDeletionPolicy.java:
```diff
@@ -70,18 +78,53 @@ public void onInit(List<? extends IndexCommit> commits) throws IOException {
+     * Releases an index commit that acquired by {@link #acquireIndexCommit(boolean)}.
+     */
+    synchronized void releaseCommit(final IndexCommit releasingCommit) {
+        assert snapshottedCommits.containsKey(releasingCommit) : "Release non-snapshotted commit;" +
```
how does that work? releasingCommit is a SnapshotIndexCommit ?

dnhatn replied:
The snapshotting commits are stored as keys in a HashMap. Both `SnapshotIndexCommit` and regular index commit inherit `equals` and `hashCode` from the root `IndexCommit`, thus they are interchangeable. This can be problematic if a regular index commit overrides `equals` or `hashCode`. 

I see some options to avoid this.

1. Exposes `SnapshotIndexCommit` to package level; makes acquireCommit and releaseCommit with `SnapshotIndexCommit` type.
2. Delegate `hashCode` and `equals` of `SnapshotIndexCommit` to the original index commit. 

WDYT?

bleskes replied:
I see. I think it's risky as the underlying IndexCommit may have a different implementation (it's not final). I see a 3rd option - we could use an identity map and sure people only release index commits they got from us. Until we need the ability to work all kind of crazyness like wrapped IndexCommits, I prefer to keep things strict.

ywelsch replied:
Lucene's SnapshotDeletionPolicy identifies the IndexCommit's based on their generation (long field).

dnhatn replied:
We discussed and agreed to keep the current implementation.

bleskes commented on core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java:
```diff
@@ -1644,20 +1643,16 @@ public void forceMerge(final boolean flush, int maxNumSegments, boolean onlyExpu
-            logger.trace("pulling snapshot");
-            return new IndexCommitRef(snapshotDeletionPolicy);
-        } catch (IOException e) {
-            throw new SnapshotFailedEngineException(shardId, e);
```
This means a potential change to the exception type. Can you double check it's OK?

dnhatn replied:
This should be ok. The method `snapshot` of `Lucene's SnapshotDeletionPolicy` throws IOException but we don't. Acquiring a commit is just increasing refCount of a commit.

bleskes commented on core/src/test/java/org/elasticsearch/index/engine/CombinedDeletionPolicyTests.java:
```diff
@@ -79,40 +77,46 @@ public void testKeepCommitsAfterGlobalCheckpoint() throws Exception {
+        final IndexCommit ref2 = indexPolicy.acquireIndexCommit(false);
+        assertThat(ref2, equalTo(c2));
+        expectThrows(UnsupportedOperationException.class, ref2::delete);
+        assertThat(translogPolicy.getMinTranslogGenerationForRecovery(), lessThanOrEqualTo(100L));
```
shouldn't this be exactly  translogGen1?

dnhatn replied:
lol. I fixed

bleskes commented on core/src/test/java/org/elasticsearch/index/engine/CombinedDeletionPolicyTests.java:
```diff
@@ -79,40 +77,46 @@ public void testKeepCommitsAfterGlobalCheckpoint() throws Exception {
+        assertThat(ref3, equalTo(c2));
+        assertThat(translogPolicy.getMinTranslogGenerationForRecovery(), equalTo(translogGen1));
+        indexPolicy.releaseCommit(ref1); // release acquired commit releases translog and commit
+        indexPolicy.onCommit(Arrays.asList(c1, c2)); // Flush new commit deletes c1
```
can we also release c2 and see that it is not deleted? I would also appreciate some randomness here - i.e., a series of commits, choose snapshot at random times. Release at random order and check that all is OK.

bleskes commented:
Thx.

On Wed, Jan 10, 2018 at 4:09 PM, Nhat Nguyen <notifications@github.com>
wrote:

> *@dnhatn* commented on this pull request.
> ------------------------------
>
> In core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java
> <https://github.com/elastic/elasticsearch/pull/28038#discussion_r160703762>
> :
>
> >          // we have to flush outside of the readlock otherwise we might have a problem upgrading
>          // the to a write lock when we fail the engine in this operation
>          if (flushFirst) {
>              logger.trace("start flush for snapshot");
>              flush(false, true);
>              logger.trace("finish flush for snapshot");
>          }
> -        try (ReleasableLock lock = readLock.acquire()) {
> -            logger.trace("pulling snapshot");
> -            return new IndexCommitRef(snapshotDeletionPolicy);
> -        } catch (IOException e) {
> -            throw new SnapshotFailedEngineException(shardId, e);
>
> This should be ok. The method snapshot of Lucene's SnapshotDeletionPolicy
> throws IOException but we don't. Acquiring a commit is just increasing
> refCount of a commit.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/elastic/elasticsearch/pull/28038#discussion_r160703762>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA9bJ8XVLnNrL1EiZiS2HJctvgoiFn2uks5tJNKmgaJpZM4RP4GN>
> .
>

ywelsch commented on core/src/main/java/org/elasticsearch/index/engine/CombinedDeletionPolicy.java:
```diff
@@ -42,12 +46,16 @@
     private final EngineConfig.OpenMode openMode;
     private final LongSupplier globalCheckpointSupplier;
+    private final ObjectIntHashMap<IndexCommit> snapshottedCommits; // Number of snapshots held against each commit point.
+    private IndexCommit safeCommit; // the most recent safe commit point - its max_seqno at most the persisted global checkpoint.
```
The safe commit point is a constantly moving target (as the global checkpoint keeps going up and new commits are being added). I wonder if it's nicer to calculate the safe commit point when it's accessed in `acquireIndexCommit`, based on the then current globalcheckpoint and the current list of commits (This will require storing the last seen indexCommits, but that would be equivalent to what's being done in Lucene's SnapshotDeletionPolicy).

bleskes replied:
@ywelsch Nhat has a follow up to trim unneeded commits as soon as the global checkpoint advances enough. This will have a side effect you mention (it will update things) with some added value.

ywelsch replied:
yes, but that involves some extra machinery to update safeCommit at the right points in time (e.g. when gcp advances)? My suggestion here makes that unnecessary?

ywelsch replied:
Ok, I think the added value of the approach in the follow-up is that the clean-up logic is also possibly more eagerly invoked (when gcp advances).

dnhatn commented:
@bleskes 

1. I've added a random test and updated the index policy not to retain translog of the snapshotting commits.
2. I realized that a delegating `equals` does not fully conform the equals contract. I decided to cast a releasing commit to the inner class. @ywelsch pointed out that Lucene's snapshot policy uses a commit-generation as a key (we are using generation and directory). I am fine with the suggestion, WDYT?

bleskes commented on core/src/main/java/org/elasticsearch/index/engine/CombinedDeletionPolicy.java:
```diff
@@ -70,18 +77,22 @@ public void onInit(List<? extends IndexCommit> commits) throws IOException {
-    public void onCommit(List<? extends IndexCommit> commits) throws IOException {
+    public synchronized void onCommit(List<? extends IndexCommit> commits) throws IOException {
         final int keptPosition = indexOfKeptCommits(commits, globalCheckpointSupplier.getAsLong());
+        lastCommit = commits.get(commits.size() - 1);
```
can we assert that the translog gen is in this commit is lower than all the ones in higher commits?

dnhatn commented:
Thanks @bleskes and @ywelsch for reviewing.
